File: Android/Paging-CRUD/app/src/main/java/com/hhp227/paging_crud/data/PostPagingSource.kt
package com.hhp227.paging_crud.data

import androidx.paging.PagingSource
import androidx.paging.PagingState
import com.hhp227.paging_crud.api.PostService
import com.hhp227.paging_crud.model.ListItem
import retrofit2.HttpException
import java.io.IOException
import kotlin.math.max

class PostPagingSource(
    private val postService: PostService,
    private val postDao: PostDao,
    private val groupId: Int
) : PagingSource<Int, ListItem.Post>() {
    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, ListItem.Post> {
        return try {
            val offset: Int = params.key ?: 0
            val loadSize: Int = params.loadSize
            val key = max(0, offset + postDao.getCount(groupId))
            val nextKey = key + loadSize
            val prevKey = key - loadSize
            val response = postService.getPostList(groupId, key, loadSize)

            if (offset == 0) postDao.deleteAll(groupId)
            if (!response.error) {
                val data = response.data ?: emptyList()

                postDao.insertAll(groupId, data)
                LoadResult.Page(
                    data = postDao.getPostList(groupId, key, nextKey),
                    prevKey = if (offset == 0) null else prevKey,
                    nextKey = if (data.isEmpty()) null else nextKey
                )
            } else {
                LoadResult.Error(
                    Throwable(response.message)
                )
            }
        } catch (e: IOException) {
            LoadResult.Error(e)
        } catch (e: HttpException) {
            LoadResult.Error(e)
        }
    }

    override fun getRefreshKey(state: PagingState<Int, ListItem.Post>): Int? {
        return state.anchorPosition?.let { anchorPosition ->
            if (postDao.isCacheEmpty(groupId)) null
            else state.closestPageToPosition(anchorPosition)?.prevKey
        }
    }
}

File: iOS/Paging-CRUD/Paging-CRUD/Data/PostPagingSource.swift
//
//  PostPagingSource.swift
//  Paging-CRUD
//

import Foundation
import Paging

final class PostPagingSource: PagingSource<Int, ListItem.Post> {
    private let postService: PostService

    private let postDao: PostDao

    private let groupId: Int

    init(postService: PostService, postDao: PostDao, groupId: Int) {
        self.postService = postService
        self.postDao = postDao
        self.groupId = groupId
        super.init()
    }

    override func load(params: LoadParams<Int>) async -> LoadResult<Int, ListItem.Post> {
        do {
            let offset = params.getKey() ?? 0
            let loadSize = PostRepository.loadSize
            let key = max(0, offset + postDao.getCount(groupId))
            let nextKey = key + loadSize
            let prevKey = key - loadSize
            let response = try await postService.getPostList(groupId: groupId, offset: key, loadSize: loadSize)

            if offset == 0 {
                postDao.deleteAll(groupId)
            }
            if !response.error {
                let data = response.data ?? []

                postDao.insertAll(groupId, data)
                return LoadResult<Int, ListItem.Post>.Page(
                    data: postDao.getPostList(groupId, key, nextKey),
                    prevKey: offset == 0 ? nil : prevKey,
                    nextKey: data.isEmpty ? nil : nextKey
                )
            } else {
                return LoadResult<Int, ListItem.Post>.Error(
                    error: NSError(
                        domain: "PostPagingSource",
                        code: 0,
                        userInfo: [NSLocalizedDescriptionKey: response.message ?? "An unexpected error occured"]
                    )
                )
            }
        } catch {
            return LoadResult<Int, ListItem.Post>.Error(error: error)
        }
    }

    override func getRefreshKey(state: PagingState<Int, ListItem.Post>) -> Int? {
        guard let anchorPosition = state.anchorPosition, !postDao.isCacheEmpty(groupId) else {
            return nil
        }
        return state.closestPageToPosition(anchorPosition)?.prevKey
    }
}
